torch/_inductor/codegen/cpp.py

```diff
@@ -611,20 +612,31 @@ def get_fx_node(self):
         return self.current_node
 
 
+def decltype_promoted(*args):
+    assert not any(
+        isinstance(arg, CppCSEVariable) and arg.is_vec for arg in args
+    ), "Promotion of vector types is not supported"
+
+    if (dt := get_promote_dtype(args)) is not None:
+        return DTYPE_TO_CPP[dt]
+    else:
+        return f"decltype({args[0]})"
+
+
 class CppOverrides(OpOverrides):
     """Map element-wise ops to C++"""
 
     @staticmethod
     def add(a, b):
-        return f"decltype({a})({a} + {b})"
+        return f"{decltype_promoted(a, b)}({a} + {b})"
```

Comment:
Is this an issue found with original implementation?

Reply:
Taking an example where `a` is of int16 and `b` is of int32. In previous implementation, the result is cast to int16 while `DtypePropagationOpsHandler` thinks it should be int32. As this PR adds some `static_assert` based type checking, this inconsistency becomes a problem necessary to be solved.